bug in question is fixed. it leads us to the next one :).

When initiating chat and telling it to call the nested tool I get this error:
org.springframework.ai.tool.execution.ToolExecutionException: ExecutionException{entityClass=interface com.bytechef.component.definition.ClusterElementDefinition, errorKey=104, errorMessageArguments=null, inputParameters=ParameterMapImpl{map={toolDescription=helps find hubspot tools, conversationId=n8js0fhtb2po7z05u7sfnsx1ervo07vj, response={responseFormat=TEXT}, userPrompt=Find HubSpot user with ID 757086160081 and return their details (name, email, role, status, and teams if available)., format=SIMPLE, toolName=hubspotTools}}} com.bytechef.exception.ExecutionException: Unsupported cluster element type: com.bytechef.component.ai.agent.tool.AiAgentChatTool$$Lambda/0x000000006f508460
	at org.springframework.ai.tool.function.FunctionToolCallback.callMethod(FunctionToolCallback.java:120)
	at org.springframework.ai.tool.function.FunctionToolCallback.call(FunctionToolCallback.java:105)
	at com.bytechef.component.ai.agent.action.AbstractAiAgentChatAction$2.lambda$call$1(AbstractAiAgentChatAction.java:387)
	at com.bytechef.component.ai.agent.action.AbstractAiAgentChatAction$2.observeAndCall(AbstractAiAgentChatAction.java:406)
	at com.bytechef.component.ai.agent.action.AbstractAiAgentChatAction$2.call(AbstractAiAgentChatAction.java:387)
	at org.springframework.ai.tool.augment.AugmentedToolCallback.call(AugmentedToolCallback.java:112)
	at org.springframework.ai.model.tool.DefaultToolCallingManager.lambda$executeToolCall$3(DefaultToolCallingManager.java:228)
	at io.micrometer.observation.Observation.observe(Observation.java:634)
Caused by: ExecutionException{entityClass=interface com.bytechef.component.definition.ClusterElementDefinition, errorKey=104, errorMessageArguments=null, inputParameters={toolDescription=helps find hubspot tools, conversationId=n8js0fhtb2po7z05u7sfnsx1ervo07vj, response={responseFormat=TEXT}, userPrompt=Find HubSpot user with ID 757086160081 and return their details (name, email, role, status, and teams if available)., format=SIMPLE, toolName=hubspotTools}} com.bytechef.exception.ExecutionException: ExecutionException{entityClass=interface com.bytechef.component.definition.ClusterElementDefinition, errorKey=104, errorMessageArguments=null, inputParameters=ParameterMapImpl{map={toolDescription=helps find hubspot tools, conversationId=n8js0fhtb2po7z05u7sfnsx1ervo07vj, response={responseFormat=TEXT}, userPrompt=Find HubSpot user with ID 757086160081 and return their details (name, email, role, status, and teams if available)., format=SIMPLE, toolName=hubspotTools}}} com.bytechef.exception.ExecutionException: Unsupported cluster element type: com.bytechef.component.ai.agent.tool.AiAgentChatTool$$Lambda/0x000000006f508460
	at com.bytechef.platform.component.service.ClusterElementDefinitionServiceImpl.doExecuteTool(ClusterElementDefinitionServiceImpl.java:462)
	at com.bytechef.platform.component.service.ClusterElementDefinitionServiceImpl.executeTool(ClusterElementDefinitionServiceImpl.java:204)
Caused by: ExecutionException{entityClass=interface com.bytechef.component.definition.ClusterElementDefinition, errorKey=104, errorMessageArguments=null, inputParameters=ParameterMapImpl{map={toolDescription=helps find hubspot tools, conversationId=n8js0fhtb2po7z05u7sfnsx1ervo07vj, response={responseFormat=TEXT}, userPrompt=Find HubSpot user with ID 757086160081 and return their details (name, email, role, status, and teams if available)., format=SIMPLE, toolName=hubspotTools}}} com.bytechef.exception.ExecutionException: Unsupported cluster element type: com.bytechef.component.ai.agent.tool.AiAgentChatTool$$Lambda/0x000000006f508460
	at com.bytechef.platform.component.service.ClusterElementDefinitionServiceImpl.doExecuteTool(ClusterElementDefinitionServiceImpl.java:455)

--- a/server/ee/libs/platform/platform-component/platform-component-remote-client/src/main/java/com/bytechef/ee/platform/component/remote/client/service/RemoteClusterElementDefinitionServiceClient.java
+++ b/server/ee/libs/platform/platform-component/platform-component-remote-client/src/main/java/com/bytechef/ee/platform/component/remote/client/service/RemoteClusterElementDefinitionServiceClient.java
@@ -82,6 +82,14 @@ public class RemoteClusterElementDefinitionServiceClient implements ClusterEleme
     }
 
     @Override
+    public Object executeTool(
+        String componentName, int componentVersion, String clusterElementName, Map<String, ?> inputParameters,
+        Map<String, ?> extensions, Map<String, ComponentConnection> componentConnections, boolean editorEnvironment) {
+
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
     public String executeWorkflowNodeDescription(
         String componentName, int componentVersion, String clusterElementName, Map<String, ?> inputParameters) {
 

--- a/server/libs/modules/components/ai/agent/src/main/java/com/bytechef/component/ai/agent/action/AbstractAiAgentChatAction.java
+++ b/server/libs/modules/components/ai/agent/src/main/java/com/bytechef/component/ai/agent/action/AbstractAiAgentChatAction.java
@@ -39,6 +39,7 @@ import com.bytechef.platform.component.definition.ParametersFactory;
 import com.bytechef.platform.component.definition.ai.agent.ChatMemoryFunction;
 import com.bytechef.platform.component.definition.ai.agent.GuardrailsFunction;
 import com.bytechef.platform.component.definition.ai.agent.ModelFunction;
+import com.bytechef.platform.component.definition.ai.agent.MultipleConnectionsToolFunction;
 import com.bytechef.platform.component.definition.ai.agent.RagFunction;
 import com.bytechef.platform.component.definition.ai.agent.ToolCallbackProviderFunction;
 import com.bytechef.platform.component.service.ClusterElementDefinitionService;
@@ -249,6 +250,9 @@ public abstract class AbstractAiAgentChatAction {
                 } catch (Exception exception) {
                     throw new RuntimeException(exception);
                 }
+            } else if (clusterElementFunction instanceof MultipleConnectionsToolFunction) {
+                toolCallbacks.add(
+                    aiAgentToolFacade.getFunctionToolCallback(clusterElement, connectionParameters, editorEnvironment));
             } else {
                 ComponentConnection componentConnection = connectionParameters.get(
                     clusterElement.getWorkflowNodeName());

--- a/server/libs/modules/components/ai/agent/src/main/java/com/bytechef/component/ai/agent/facade/AiAgentToolFacade.java
+++ b/server/libs/modules/components/ai/agent/src/main/java/com/bytechef/component/ai/agent/facade/AiAgentToolFacade.java
@@ -86,6 +86,42 @@ public class AiAgentToolFacade extends AbstractToolFacade {
         return builder.build();
     }
 
+    public ToolCallback getFunctionToolCallback(
+        ClusterElement clusterElement, Map<String, ComponentConnection> componentConnections,
+        boolean editorEnvironment) {
+
+        ClusterElementDefinition clusterElementDefinition =
+            clusterElementDefinitionService.getClusterElementDefinition(
+                clusterElement.getComponentName(), clusterElement.getComponentVersion(),
+                clusterElement.getClusterElementName());
+
+        Map<String, ?> toolParameters = clusterElement.getParameters();
+
+        List<FromAiResult> fromAiResults = extractFromAiResults(toolParameters);
+
+        FunctionToolCallback.Builder<Map<String, Object>, Object> builder = FunctionToolCallback.builder(
+            getToolName(clusterElementDefinition.getComponentName(), clusterElementDefinition.getName(),
+                toolParameters),
+            getMultipleConnectionsToolCallbackFunction(
+                clusterElement.getComponentName(), clusterElement.getComponentVersion(),
+                clusterElementDefinition.getName(), toolParameters, clusterElement.getExtensions(),
+                componentConnections, editorEnvironment))
+            .inputType(Map.class)
+            .inputSchema(FromAiInputSchemaUtils.generateInputSchema(fromAiResults));
+
+        String toolDescription = getToolDescription(toolParameters, clusterElement.getExtensions());
+
+        if (toolDescription == null) {
+            toolDescription = clusterElementDefinition.getDescription();
+        }
+
+        if (toolDescription != null) {
+            builder.description(toolDescription);
+        }
+
+        return builder.build();
+    }
+
     private Function<Map<String, Object>, Object> getFromAiToolCallbackFunction(
         String componentName, int componentVersion, String clusterElementName, Map<String, ?> parameters,
         @Nullable ComponentConnection componentConnection, boolean editorEnvironment) {
@@ -100,6 +136,23 @@ public class AiAgentToolFacade extends AbstractToolFacade {
             return clusterElementDefinitionService.executeTool(
                 componentName, componentVersion, clusterElementName, MapUtils.concat(request, resolvedParameters),
                 componentConnection, editorEnvironment);
+        };
+    }
+
+    private Function<Map<String, Object>, Object> getMultipleConnectionsToolCallbackFunction(
+        String componentName, int componentVersion, String clusterElementName, Map<String, ?> parameters,
+        Map<String, ?> extensions, Map<String, ComponentConnection> componentConnections, boolean editorEnvironment) {
+
+        return request -> {
+            Map<String, Object> resolvedParameters = new HashMap<>();
+
+            for (Map.Entry<String, ?> entry : parameters.entrySet()) {
+                resolvedParameters.put(entry.getKey(), resolveParameterValue(entry.getValue(), request));
+            }
+
+            return clusterElementDefinitionService.executeTool(
+                componentName, componentVersion, clusterElementName, MapUtils.concat(request, resolvedParameters),
+                extensions, componentConnections, editorEnvironment);
         };
     }
 

--- a/server/libs/platform/platform-component/platform-component-api/src/main/java/com/bytechef/platform/component/service/ClusterElementDefinitionService.java
+++ b/server/libs/platform/platform-component/platform-component-api/src/main/java/com/bytechef/platform/component/service/ClusterElementDefinitionService.java
@@ -61,6 +61,10 @@ public interface ClusterElementDefinitionService extends OperationDefinitionServ
         String componentName, int componentVersion, String clusterElementName, Map<String, ?> inputParameters,
         @Nullable ComponentConnection componentConnection, boolean editorEnvironment);
 
+    Object executeTool(
+        String componentName, int componentVersion, String clusterElementName, Map<String, ?> inputParameters,
+        Map<String, ?> extensions, Map<String, ComponentConnection> componentConnections, boolean editorEnvironment);
+
     String executeWorkflowNodeDescription(
         String componentName, int componentVersion, String clusterElementName, Map<String, ?> inputParameters);
 

--- a/server/libs/platform/platform-component/platform-component-service/src/main/java/com/bytechef/platform/component/service/ClusterElementDefinitionServiceImpl.java
+++ b/server/libs/platform/platform-component/platform-component-service/src/main/java/com/bytechef/platform/component/service/ClusterElementDefinitionServiceImpl.java
@@ -46,6 +46,7 @@ import com.bytechef.platform.component.definition.ActionContextAdapater;
 import com.bytechef.platform.component.definition.ClusterRootComponentDefinition;
 import com.bytechef.platform.component.definition.ParametersFactory;
 import com.bytechef.platform.component.definition.PropertyFactory;
+import com.bytechef.platform.component.definition.ai.agent.MultipleConnectionsToolFunction;
 import com.bytechef.platform.component.definition.ai.agent.ToolCallbackProviderFunction;
 import com.bytechef.platform.component.definition.datastream.ClusterElementResolverFunction;
 import com.bytechef.platform.component.domain.ClusterElementDefinition;
@@ -203,6 +204,24 @@ public class ClusterElementDefinitionServiceImpl implements ClusterElementDefini
 
         return doExecuteTool(
             componentName, componentVersion, clusterElementName, inputParameters, componentConnection,
+            clusterElementContext);
+    }
+
+    @Override
+    public Object executeTool(
+        String componentName, int componentVersion, String clusterElementName, Map<String, ?> inputParameters,
+        Map<String, ?> extensions, Map<String, ComponentConnection> componentConnections, boolean editorEnvironment) {
+
+        ComponentConnection firstConnection = componentConnections.isEmpty()
+            ? null : componentConnections.values()
+                .iterator()
+                .next();
+
+        ClusterElementContext clusterElementContext = contextFactory.createClusterElementContext(
+            componentName, componentVersion, clusterElementName, firstConnection, editorEnvironment);
+
+        return doExecuteTool(
+            componentName, componentVersion, clusterElementName, inputParameters, extensions, componentConnections,
             clusterElementContext);
     }
 
@@ -442,6 +461,45 @@ public class ClusterElementDefinitionServiceImpl implements ClusterElementDefini
         Parameters connectionParameters = ParametersFactory.create(componentConnection);
 
         try {
+            if (clusterElement instanceof ToolCallbackProviderFunction toolCallbackProviderFunction) {
+                return toolCallbackProviderFunction.apply(inputParameters, connectionParameters, context);
+            }
+
+            if (clusterElement instanceof ToolFunction toolFunction) {
+                return toolFunction.apply(inputParameters, connectionParameters, context);
+            }
+
+            throw new ExecutionException(
+                "Unsupported cluster element type: " + clusterElement.getClass()
+                    .getName(),
+                inputParameters, ClusterElementDefinitionErrorType.EXECUTE_PERFORM);
+        } catch (Exception exception) {
+            if (exception instanceof ProviderException) {
+                throw (ProviderException) exception;
+            }
+
+            throw new ExecutionException(
+                exception, inputParameterMap, ClusterElementDefinitionErrorType.EXECUTE_PERFORM);
+        }
+    }
+
+    private Object doExecuteTool(
+        String componentName, Integer componentVersion, String clusterElementName, Map<String, ?> inputParameterMap,
+        Map<String, ?> extensionMap, Map<String, ComponentConnection> componentConnections,
+        ClusterElementContext context) {
+
+        Object clusterElement = getClusterElement(componentName, componentVersion, clusterElementName);
+
+        Parameters inputParameters = ParametersFactory.create(inputParameterMap);
+        Parameters connectionParameters = ParametersFactory.create(Map.of());
+        Parameters extensions = ParametersFactory.create(extensionMap);
+
+        try {
+            if (clusterElement instanceof MultipleConnectionsToolFunction multipleConnectionsToolFunction) {
+                return multipleConnectionsToolFunction.apply(
+                    inputParameters, connectionParameters, extensions, componentConnections, context);
+            }
+
             if (clusterElement instanceof ToolCallbackProviderFunction toolCallbackProviderFunction) {
                 return toolCallbackProviderFunction.apply(inputParameters, connectionParameters, context);
             }

